megamek/src/megamek/client/bot/princess/FireControl.java

```diff
@@ -1352,16 +1352,30 @@ protected double calcCommandUtility(final Targetable target) {
     }
 
     private boolean isCommander(final Entity entity) {
-        return entity.isCommander() || entity.hasC3M() || entity.hasC3i() || entity.hasC3MM() ||
-               (owner.getHighestEnemyInitiativeId() == entity.getId());
+        if (owner.getFireControlState().commanderCached(entity)) {
+            return owner.getFireControlState().isCommander(entity);
+        } else {
+            owner.getFireControlState().setCommander(entity, 
+                    entity.isCommander() || entity.hasC3M() || entity.hasC3i() || entity.hasC3MM() ||
+                    (owner.getHighestEnemyInitiativeId() == entity.getId()));
+            
+            return owner.getFireControlState().isCommander(entity);
+        }
     }
 
     private boolean isSubCommander(final Entity entity) {
-        final int initBonus = entity.getHQIniBonus() + entity.getQuirkIniBonus();  //removed in IO + entity.getMDIniBonus()
-        return entity.hasC3() || entity.hasTAG() || entity.hasBoostedC3() || entity.hasNovaCEWS() ||
-               entity.isUsingSpotlight() || entity.hasBAP() || entity.hasActiveECM() || entity.hasActiveECCM() ||
-               entity.hasQuirk(OptionsConstants.QUIRK_POS_IMPROVED_SENSORS) || entity.hasEiCockpit() ||
-               (0 < initBonus);
+        if (owner.getFireControlState().subCommanderCached(entity)) {
+            return owner.getFireControlState().isSubCommander(entity);
```

Comment:
This can be pushed outside of the else, so you instead just check if it is not cached and add it if that is the case. Alternatively, setSubCommander could return the boolean.